### ISSUE TITLE
FatserRCNN detection of objects in new image not good. mAP value is .48 of custom dataset evaluation 

### ISSUE THREAD

**Issue by ajaysharmahcl** [Microsoft/CNTK#2766]
Hello Team,

I have used FasterRCNN to train/evaluate AlexNet caffe model on custom vehicle dataset. Downloaded around 150 images with different objects and annotated with 4 categories using VOTT tool. Result of evaluation is mAP=.48 . when i used detection method to know objects in a new fresh image , it results only detecting it one object most of times. Some time it tags correctly like car and person but not all the time. Is there any set of parameters that need to be set to get the good detection result please? 
Could anyone please suggest which are important parameters with values those play important role in detection mechanism with good result (mAP close to 100) ?
below are the parameters values i have set in fasterrcnn config file -
# Learning parameters
__C.CNTK.L2_REG_WEIGHT = 0.0005
__C.CNTK.MOMENTUM_PER_MB = 0.9
# The learning rate multiplier for all bias weights
__C.CNTK.BIAS_LR_MULT = 2.0

# E2E learning parameters
__C.CNTK.E2E_MAX_EPOCHS = 20
__C.CNTK.E2E_LR_PER_SAMPLE = [0.001] * 10 + [0.0001] * 10 + [0.00001]

# 4-stage learning parameters (alternating training scheme)
__C.CNTK.RPN_EPOCHS = 16
__C.CNTK.RPN_LR_PER_SAMPLE = [0.001] * 12 + [0.0001] * 4
__C.CNTK.FRCN_EPOCHS = 8
__C.CNTK.FRCN_LR_PER_SAMPLE = [0.001] * 6 + [0.0001] * 2

# Maximum number of ground truth annotations per image
__C.INPUT_ROIS_PER_IMAGE = 1500 # changed from 50 to 1000
__C.IMAGE_WIDTH = 850
__C.IMAGE_HEIGHT = 850

# Sigma parameter for smooth L1 loss in the RPN and the detector (DET)
__C.SIGMA_RPN_L1 = 3.0
__C.SIGMA_DET_L1 = 1.0

# NMS threshold used to discard overlapping predicted bounding boxes
__C.RESULTS_NMS_THRESHOLD = 0.2
# all bounding boxes with a score lower than this threshold will be considered background
__C.RESULTS_NMS_CONF_THRESHOLD = 0.0

# Enable plotting of results generally / also plot background boxes / also plot unregressed boxes
__C.VISUALIZE_RESULTS = False
__C.DRAW_NEGATIVE_ROIS = False
__C.DRAW_UNREGRESSED_ROIS = False
# only for plotting results: boxes with a score lower than this threshold will be considered background
__C.RESULTS_BGR_PLOT_THRESHOLD = 0.1

#
# Training parameters
#

__C.TRAIN = edict()

# Use horizontally-flipped images during training?
__C.TRAIN.USE_FLIPPED = True
# If set to 'True' conv layers weights from the base model will be trained, too
__C.TRAIN_CONV_LAYERS = True

# RPN parameters
# IOU >= thresh: positive example
__C.TRAIN.RPN_POSITIVE_OVERLAP = 0.7
# IOU < thresh: negative example
__C.TRAIN.RPN_NEGATIVE_OVERLAP = 0.3
# If an anchor statisfied by positive and negative conditions set to negative
__C.TRAIN.RPN_CLOBBER_POSITIVES = False
# Max number of foreground examples
__C.TRAIN.RPN_FG_FRACTION = 0.5
# Total number of examples
__C.TRAIN.RPN_BATCHSIZE = 256
# NMS threshold used on RPN proposals
__C.TRAIN.RPN_NMS_THRESH = 0.7
# Number of top scoring boxes to keep before apply NMS to RPN proposals
__C.TRAIN.RPN_PRE_NMS_TOP_N = 12000
# Number of top scoring boxes to keep after applying NMS to RPN proposals
__C.TRAIN.RPN_POST_NMS_TOP_N = 2000
# Proposal height and width both need to be greater than RPN_MIN_SIZE (at orig image scale)
__C.TRAIN.RPN_MIN_SIZE = 16

# Detector parameters
# Minibatch size (number of regions of interest [ROIs]) -- was: __C.TRAIN.BATCH_SIZE = 128
__C.NUM_ROI_PROPOSALS = 128
# Fraction of minibatch that is labeled foreground (i.e. class > 0)
__C.TRAIN.FG_FRACTION = 0.25
# Overlap threshold for a ROI to be considered foreground (if >= FG_THRESH)
__C.TRAIN.FG_THRESH = 0.5
# Overlap threshold for a ROI to be considered background (class = 0 if
# overlap in [LO, HI))
__C.TRAIN.BG_THRESH_HI = 0.5
__C.TRAIN.BG_THRESH_LO = 0.0

# Normalize the targets using "precomputed" (or made up) means and stdevs
__C.BBOX_NORMALIZE_TARGETS = True
__C.BBOX_NORMALIZE_MEANS = (0.0, 0.0, 0.0, 0.0)
__C.BBOX_NORMALIZE_STDS = (0.1, 0.1, 0.2, 0.2)

#
# Testing parameters
#

__C.TEST = edict()

## NMS threshold used on RPN proposals
__C.TEST.RPN_NMS_THRESH = 0.7
## Number of top scoring boxes to keep before apply NMS to RPN proposals
__C.TEST.RPN_PRE_NMS_TOP_N = 6000
## Number of top scoring boxes to keep after applying NMS to RPN proposals
__C.TEST.RPN_POST_NMS_TOP_N = 300
# Proposal height and width both need to be greater than RPN_MIN_SIZE (at orig image scale)
__C.TEST.RPN_MIN_SIZE = 16

**Comment by Dozer3D** [comment]
Not sure if you are saying you trained the network with only 150 images? or just used 150 to test the evaluation... I get really good results for vehicle detection using the Faster-rcnn CNTK, but I created my own training set consisting of 7500 images (painfully downloaded all of them myself over several days...) and then retrained the network following the instructions in CNTKs Faster-rcnnn tutorial. I left all parameters that you listed above exactly the same as the CNTK samples - didn't change any of them. If you are only use 150 for training, I say you'd need a lot more to improve the accuracy.

**Comment by ajaysharmahcl** [comment]
Thanks a lot Dozer for quick response. Yes i tool only set of 150 images in train folder while around 40 images in testimage folder. I annotated them using VOTT tool. I understand that more number of images in trained dataset mean more accuracy but it's time consuming process. Even when annotating them there is no guarantee that we are scaling then with right ROI & positions. Is there any pre-trained dataset available for vehicles that we can take as baseline and add few more categories and images to it if required?
Also please advise on ImageNet dataset available in CNTK Datasets directory ? I will appreciate if following questions can be addressed -
1. Can we use ImageNet dataset for the same (I am using CPU enabled machine). If yes then why pyhton script to download the dataset is missing ?
2. Is ImageNet can be trained on CPU enabled machine ? if yes then i think annotation not required because model will be already trained with images..Please correct if understanding is not right
3. Is there any other tool to annotate the images to create custom data set ?
4. any new way updated in CNTK that produce more accuracy and better detection result ?

Many Thanks in advance !  to this whole community to help people sorting out the issues.

**Comment by Dozer3D** [comment]
NVIDIA's Digits software had something similar for vehicle detection, but it wasn't any near as good as Faster rcnn. However, you might be able to use the same Kitti dataset. But you'd still have to write something that translates their file format containg the annotations to one compatible with the CNTK sample.
[https://github.com/NVIDIA/DIGITS/tree/master/examples/object-detection](url)

**Comment by ajaysharmahcl** [comment]
Thanks Dozer again. Is it required GPU enabled machine as includes NVIDIA?
Also what is correct process to train model on own dataset ? any other way to annotate the images excpet VOTT tool?

**Comment by Dozer3D** [comment]
I was not clear in my last post. If you want vehicle data, you can use this [Kitti dataset](http://www.cvlibs.net/datasets/kitti/), but I could not use this myself because of the non commercial license.

> Also what is correct process to train model on own dataset ?

I followed the instructions here: [Run Faster R-CNN on your own data](https://docs.microsoft.com/en-us/cognitive-toolkit/Object-Detection-using-Faster-R-CNN#run-faster-r-cnn-on-your-own-data)

> any other way to annotate the images excpet VOTT tool?

I wrote my own tool for that because I wanted a detailed understanding off how everything works.
You have probably seen this link [Prepare a custom dataset ](https://docs.microsoft.com/en-us/cognitive-toolkit/Object-Detection-using-Fast-R-CNN#train-on-your-own-data) which contains 2 options:

1) Option 1: Visual Object Tagging Tool (Recommended)
2) Option 2: Using Annotation Scripts

If you want to write your own tool, then you need to understand how to write to the following files:
class_map.txt, test_img_file.txt, test_roi_file.txt, train_img_file.txt, train_roi_file.txt. So my tool maintains a database (folder) to store its images and the ROI info etc, and then has an Export button that will then create the previously mentioned files and populate the  "negative", "postive" and "testimages" folders with the images.

**Comment by ajaysharmahcl** [comment]
Thanks a lot again Dozer. Really appreciated.
This looks like kitti dataset has dependency on GPU/NVDIA and associated with license T&C too before we use it. I cannot use it on CPU enabled machine. ? is my understanding right ?

Is there any other trained vehicle dataset available on GITHUB that can be used in CNTK detection process considering CPU machine only. ?

I have explored few articles which talk about trained dataset but in different format. I will look a way to translate those dataset into CNTK format and use.

Looks like better way turn to have a large data set of  our own images and train them using pre-trained alexnet model. 

Thank You

**Comment by Dozer3D** [comment]
> This looks like kitti dataset has dependency on GPU/NVDIA

I don't think so, the datasets are typically just large numbers of image and additional files containing ROIs etc. Typically created by universities and have licenses that prevent non commercial use. Makes no difference if you use CPU or GPU. The article I linked to before might have confused you, I was just saying that NVIDIA used the kitti datset for one of their detectnet sample in their DIGITS software - but forget that.

> Looks like better way turn to have a large data set of our own images and train them using pre-trained alexnet model.

Thats what I did and worked really well. I found downloading images for used car sales websites worked well, but at the time I couldn't find any usable data sets either. not sure if that has changed now.

**Comment by ajaysharmahcl** [comment]
Ok. thank you Dozer. Really appreciated.

**Comment by ajaysharmahcl** [comment]
Dear Dozer,

I found car data sets with huge number of images (trained&labeled) at below Url -
http://ai.stanford.edu/~jkrause/cars/car_dataset.html

i have downloaded a dataset but not able to translate it into format that CNTK accept for evaluation and detection. It has some .mat files which run_faster_rcnn.py does not recognized.
Would you please advise how can i convert this dataset into cntk acceptable format please?
thank you in advance.
